Source/Core/Common/Image.cpp

```diff
@@ -39,23 +41,52 @@ bool LoadPNG(const std::vector<u8>& input, std::vector<u8>* data_out, u32* width
   return true;
 }
 
+static void WriteCallback(png_structp png_ptr, png_bytep data, size_t length)
+{
+  std::vector<u8>* buffer = static_cast<std::vector<u8>*>(png_get_io_ptr(png_ptr));
+  buffer->insert(buffer->end(), data, data + length);
+}
+
+bool SavePNG0(png_structp png_ptr, png_infop info_ptr, int png_format, u32 width, u32 height,
+              int level, std::vector<u8>* buffer, const std::vector<const u8*>& rows) noexcept
+{
+#ifdef _MSC_VER
+#pragma warning(push)
+// Disable warning for setjmp in C++
+#pragma warning(disable : 4611)
+#endif
+  if (setjmp(png_jmpbuf(png_ptr)) != 0)
```

Comment:
i seem to remember purposefully getting rid of setjmp usage in libpng-related code before; is it really required?

Reply:
The side effects (and reason for warning) can be pretty subtle and easy to get wrong.

Reply:
The code previously used the simplified libpng API (`png_image_write_to_memory`); as far as I can tell there is no way to set the compression level with that API.

`png_image_write_to_memory` [calls](https://github.com/dolphin-emu/dolphin/blob/890a5ed99ad4130d68b3e7228c3398445c33cdc4/Externals/libpng/pngwrite.c#L2242) `png_safe_execute`, which [uses](https://github.com/dolphin-emu/dolphin/blob/890a5ed99ad4130d68b3e7228c3398445c33cdc4/Externals/libpng/pngerror.c#L945) `setjmp` internally.  So `setjmp` was being used before, though not directly in Dolphin code.

I did try to isolate the `setjmp` call to its own function and have the C++ objects be constructed and destructed outside of that function, but I'm not sure if that's guaranteed to be safe (either by the standard or in practice due to inlining).

Reply:
Since I'm the one who ported this over to the simplified API: IMO if you want to use the complex API, we should make a separate C (not C++) compilation unit and use the `setjmp` there. Otherwise this is inviting danger for anyone who touches the code in the future.

Reply:
If there's a way to write it in C and have the write callback also be C that would be nice. It's not a huge deal to me either way but just a bit of unfortunate code pattern by libpng :/

Reply:
I've moved `SavePNG0` into a C file (which also required disabling precompiled headers for that file).

According to [cppreference on `longjmp`](https://en.cppreference.com/w/cpp/utility/program/longjmp),

> No destructors for automatic objects are called. If replacing of `std::longjmp` with `throw` and `setjmp` with `catch` would execute a non-trivial destructor for any automatic object, the behavior of such `std::longjmp` is undefined. 

Since `longjmp` isn't called by the callback, I think it's fine for the callback to be in C++ code.  I also don't want to try and port the old buffer-resizing code to C, partially because I think it'd be fiddly to do and partially because if the buffer ends up being too small, I don't want to run the compression process a second time (though that probably won't happen often for higher compression levels).

Reply:
The reason I suggested that the callback be C is because std::vector::insert can throw (which comes with it's own problems but might also cause dtors to be hit on paths you wouldn't expect). But, dolphin doesn't care about such exceptions at all anyway, so 🤷‍♂️. It's fine how it is imo but just FYI